### ISSUE TITLE
Use additional MappingSchema on Host with ServiceModelDataContext

### ISSUE THREAD

**Issue by frankiDotNet** [linq2db/linq2db#1726]
Based on [this ](https://github.com/linq2db/linq2db/issues/1642#issuecomment-472772995) request I wanted to create a mapping from different tables to one class.
This works fine with the default class that inherits from DataConnection.

But how can I do this with the ServiceModelDataContext ?

```
public partial class DbContextService : LinqToDB.ServiceModel.ServiceModelDataContext
   {

    public DbContextService(Binding binding, EndpointAddress endpointAddress) : 
    base(binding, endpointAddress) { }
}
```

the default looks like this :

```
public partial class DbContext : DataConnection
   {
      public DbContext()
      {
         RetryPolicy = new FbRetryPolicy();
      }
      public DbContext(MappingSchema ms) : this()
      {
         if (ms != null && this.MappingSchema != ms)
         {
            this.AddMappingSchema(ms);
         }
      }
}
```

How to add the MappingSchema to the ServiceModelDataContext-class?

**Comment by sdanyliv** [comment]
`ctx.MappingSchema = ms` ?

**Comment by Franki1986** [comment]
Is it the same like adding a MappingSchema?

**Comment by sdanyliv** [comment]
Not sure. You have to test it yourself.

**Comment by Franki1986** [comment]
Basically this works:

```
public partial class DbContextService : LinqToDB.ServiceModel.ServiceModelDataContext
   {

    public DbContextService(Binding binding, EndpointAddress endpointAddress) : 
    base(binding, endpointAddress) { }

    public DbContextService(Binding binding, EndpointAddress endpointAddress, MappingSchema  ms) : 
    base(binding, endpointAddress) {
        this.MappingSchema = new MappingSchema(this.MappingSchema, ms);
    }
}
```

But you have to generate the additional MappingSchema on the wcf client. 
In my case this is not ideal , because every WCF host has its own version of database (Based on different application versions).

In my situation on the WCF host the different MappingSchemas are created on start and saved in a static Dictionary.

So my idea was to pass the key to over the wcf message headers, get it on the host and add the specified mappigSchema:


WCF-Client
...
 ```
string serviceEndpointAddress = myEndpointAddress;
         AddressHeader header = AddressHeader.CreateAddressHeader("CustomMappingSchemaKey", String.Empty, "ExternalCard");
         var endpoint = new EndpointAddress(new Uri(serviceEndpointAddress), header);
```
...

WCF-host:

```
public class LinqServiceProvider : LinqService
   {
public override DataConnection CreateDataContext(string configuration)
      {
         var context = OperationContext.Current;
         var headers = context?.IncomingMessageHeaders;
         string key = null;
         if (headers != null)
         {
            key = headers.GetHeader<string>("CustomMappingSchema", String.Empty);
         }
         var defaultConfiguration = DataConnection.DefaultConfiguration;
         var connection = DataConnection.GetConnectionString(defaultConfiguration);
         var result = base.CreateDataContext(configuration ?? defaultConfiguration);
         if (key != null)
         {            result.AddMappingSchema(GetMappingSchema(key));
         }
         return result;
      }
}
```

Everything works like expected, if I debug the LinqService-class in the [ExecuteReader ](https://github.com/linq2db/linq2db/blob/8b08c92e526bc56bb5c385a9ed96b1d8e312f683/Source/LinqToDB/ServiceModel/LinqService.cs#L134) method, there is the MappingSchema with the registered Mapping, but in the end this mapping is not used! I get an Exception that the table (my Type class name) is unknown and in the logged sql string no of my mapping is used, only the class and property names of it are used.

**Comment by Franki1986** [comment]
So for clarification adding a MappingSchema on wcf host side does not work for the wcf client:

```
public class LinqServiceProvider : LinqService
   {
    public override DataConnection CreateDataContext(string configuration)
      {
         var dbContext = new MyDbContext();
         var additionalMapping = GenerateWithFluentBuilder(); // Generates a MappingSchema with fluernt builder
         dbContext.AddmappingSchema(additionalMapping);
         return dbContext;
      }
}
```

**Comment by MaceWindu** [comment]
FYI, for version 3.0 I've added host schema support as a part of #730  in PR #1512

**Comment by Franki1986** [comment]
So I have testet this with version 3.0 preview:

```
public class LinqServiceProvider : LinqService
   {
    public override DataConnection CreateDataContext(string configuration)
      {
         var dbContext = new MyDbContext();
         var additionalMapping = GenerateWithFluentBuilder(); // Generates a MappingSchema with fluernt builder
         dbContext.AddmappingSchema(additionalMapping);

        using(var dbLocal = new MyDbContext(additionalMapping)){
             var query = dbLocal.GetTable<Person>(); // returns the right sql
        }

         return dbContext;
      }
}
```


```
using(var db = new MyServiceContext()){
     var query = dbLocal.GetTable<Person>(); // returns the wrong sql
}
```


Something I missed here?

**Comment by MaceWindu** [comment]
What do you mean by "wrong sql"? Wrong results from query or incorrect SQL query text from query.ToString()? Latter never worked for remote context - always returned approximate sql from query AST.

**Comment by Franki1986** [comment]
Ah ok, thought the MappingSchema declared on host side would be  used to generate the sql string. I am using a dynamic mapping schema on each of my hosts, so basically the generated sql would differ on each server.
Do you have a little example what differs infront to the old 2.9 version?

**Comment by MaceWindu** [comment]
To clarify why it works like that: when you call ToString on query over remote context - we build SQL locally using AST ToString() implementation, which provides generic provider-agnostic SQL generally for debugging purposes. To fix it we should redirect this call to server, where it will be build using provider's sql builder. If you need real SQL on client - it will be better to create separate feature request.

For 2.x -> 3.0 migration most inportant change is the way we read data from DbReader on server.
In 2.x we just read data from DbReader and send it client, where it used to map to results. This approach was causing issue #730 where any custom logic to read data from data reader for specific type, defined for provider, were ignored and could have lead to incorrect data or errors.

To make it more clear, e.g. for Firebird [those](https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/DataProvider/Firebird/FirebirdDataProvider.cs#L35-L39) custom read methods are not called over remote context in 2.x.

While it fixed this major issue, this change is not free:
1. we need to support serialization of values, returned by custom read logic. For this I've added serialization mapping schema to server and client parts, which defines serialization logic for regular types (numbers, strings, date/time types), but if some non-standard type is used - user will need to register serialization logic for this type on server and client in form of Type<->string converters.
2. we need to know target field type to call corresponding data read method. In general this is not an issue, but I've discovered several very specific cases that produce `SELECT *` queries (no column list) and I cannot properly read data from reader for such cases. While it is major regression (such queries doesn't work over remote context in 3.0 now), it is not a big issue, as I don't expect people use such broken queries anyway. Also such queries means we have bug in linq2db itself so they will be fixed at some point (preferably before 3.0 release, but some of them will stay here for some time).

**Comment by Franki1986** [comment]
... user will need to register serialization logic for this type on server and client in form of Type<->string converters..
I had the same problem when trying to create a custom WCF-Service that generates the complete result on server side and returns it back serialized.. At the end I had to use JSON.Net for serializing also anonymous types..

**Comment by Franki1986** [comment]
Sorry I tried it another time but I have to ask to clarify:

if I return this Context from my WCF-Service:

```
public class LinqServiceProvider : LinqService
   {
    public override DataConnection CreateDataContext(string configuration)
      {
         var dbContext = new MyDbContext();
         var additionalMapping = GenerateWithFluentBuilder(); // Generates a MappingSchema with fluernt builder
         dbContext.AddmappingSchema(additionalMapping);              
         return dbContext;
      }
}
```

and call it like this:

```
using(var db = new MyServiceContext()){
    var result =  db.GetTabl<Person>().ToList();
}
```

The Person Table is not mapped, only over the FluentMappinBuilder on server side, that adds the MappingSchema to the current context. 
If I call ToList() on client side the MappingSchema that is declared on server side is not used. In my case I have a table called 'CUSTOMER' with different columns, that are dynamically mapped with the fluent builder. But on client side I get the exception that Table 'Person' does not exist, because there is no mapping to this.

**Comment by MaceWindu** [comment]
You need to register your fluent mapping in MyServiceContext too.

**Comment by MaceWindu** [comment]
CLosing as `LinqService.MappingSchema` available since v4.0 release : https://github.com/linq2db/linq2db/pull/3410
